include/ginkgo/core/solver/lower_trs.hpp

```diff
@@ -0,0 +1,279 @@
+                              _parameters_name##_type, ValueType, IndexType> { \
+        friend class ::gko::EnablePolymorphicObject<                           \
+            _factory_name,                                                     \
+            ::gko::solver::LowerTrsFactory<ValueType, IndexType>>;             \
+        friend class ::gko::enable_parameters_type<_parameters_name##_type,    \
+                                                   _factory_name>;             \
+        using ::gko::solver::EnableDefaultLowerTrsFactory<                     \
+            _factory_name, _lower_trs, _parameters_name##_type, ValueType,     \
+            IndexType>::EnableDefaultLowerTrsFactory;                          \
+    };                                                                         \
+    friend ::gko::solver::EnableDefaultLowerTrsFactory<                        \
+        _factory_name, _lower_trs, _parameters_name##_type, ValueType,         \
+        IndexType>;                                                            \
+                                                                               \
+private:                                                                       \
+    _parameters_name##_type _parameters_name##_;                               \
+                                                                               \
+public:                                                                        \
+    static_assert(true,                                                        \
+                  "This assert is used to counter the false positive extra "   \
+                  "semi-colon warnings")
+
+
+/**
+ * LOWER_TRS is the triangular solver which solves the system L x = b, when L is
+ * a lower triangular matrix. It works best when passing in a matrix in CSR
+ * format. If the matrix is not in CSR, then the generate step converts it into
+ * a CSR matrix.
+ *
+ * @note When the factory is being generated with an empty matrix so that it can
+ * be copied from another matrix, only a CSR matrix can be passed in because it
+ * does not make sense to convert an empty matrix in another format to CSR.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indices
+ *
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class LowerTrs : public EnableLinOp<LowerTrs<ValueType, IndexType>>,
+                 public Preconditionable {
+    friend class EnableLinOp<LowerTrs>;
+    friend class EnablePolymorphicObject<LowerTrs, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Gets the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
```

Comment:
Here, the return type might be useful to stay `std::shared_ptr<const LinOp>` since we do that with every class, I am really unsure about that.
But if you remove `system_matrix_` (and replace it here with `csr_system_matrix_` or just change the type of `sytem_matrix_`), this code does not need to be changed (other than potentially renaming) since it would be an implicit up-cast.

(Note: I kind of messed up the order, for this comment to make sense, read the comment 3 or 4 below this one.)